### PR TITLE
fix(web): fixes OSK-interaction issues in Float, Toolbar UI modules 

### LINKED ISSUE #6685: bug(Web): Show On Screen Keyboard is not working in Keyman Web
**Description**
On Screen Keyboard appears at the right corner of the Keyman Web bar is not working consistently. 

**To Reproduce**
1. Go to 'KeymanWeb.com(beta)' site. 
2. Click on 'Languages' drop down list. 
3. Click on 'Asia' label name. 
4. Scroll down to 'Tamil'
5. Verify that the Tamil language appears on the Keyboard bar. 
6. Click on the Text Area.
7. Verify that the OSK appears at the bottom of the Text Area. 
8. Verify that the Keyboard icon which appears at the top right corner of the Windows is enabled.
9. Click on the Close button on the OSK. 
10. Verify that the OSK disappears on the Screen. 
11. Click once again on the text area. 

Here, I noticed that OSK is missing under the Text Area. 

---
**Expected behavior**
The On Screen Keyboard should appear under the text area, after enabling the Keyboard Icon.

**Screenshots**
 

---

**KeymanWeb:**
 - OS: Windows 10, Mac OS 12.1 Monterey, Windows 11
 - Browser:  Chrome, Firefox
 - KeymanWeb Version: 15.0.252
 - URL:  https://keymanweb.com/#ta,Keyboard_basic_kbdintam
 - 
---

**Additional context**
Only reproducible in Keymanweb.com Site.

### PR DESCRIPTION
**Pull request by jahorton**
Fixes #6685.

As it turns out, both the Float and Toolbar UI modules were broken by changes made during the `InlinedOSK` work for 15.0.

1. The OSK is no longer initialized immediately, meaning the file-level variable for the `osk` had to be eliminated.  The adjustments were fine with minification disabled, but not when enabled.
    - Refer to #5412.
2. The encapsulated modeling of OSK hide vs show behavior caused the OSK to be hidden (via `Promise`-deferment) on loss of focus when users interacted with certain UI elements, as the two modules didn't make full use of `keyman.activatingUI` to note when the active element should be preserved on temporary loss of focus.
    - As the hiding operation was via `Promise`-deferment, any "show" commands would actually work... and then would instantly be overridden once the deferred operation kicked in.
3.  `osk.show()` is now _only_ an API method, as display state tracking based on page interaction is now tracked by other mannerisms.  While _hiding_ the OSK was set to update cookie state in one of its submethods... someone (`git blame` -> @jahorton) forgot to also update cookie state when setting it visible.  Oops.

For reviews, "hide whitespace" is strongly suggested.

## User Testing

<details>
<summary><strong>Desktop instructions</strong></summary>

The important part:  pay extra attention to in-browser desktop form-factor tests, as those are the most likely to break.  Make sure the UI modules load properly and allow keyboard-swapping as expected.

There is a new set of test pages designed to facilitate testing the UI modules:

image

The pages may be found through this link on the testing index page:

image

In order to test compatibility with the different UI module types, you'll need to mildly edit the unminified test page's source:

Here's the UI you should expect to see for each test:
- Button UI
    image
    - Note:  the button will likely display above the first input element, rather than beside the current text input receiver.  This is fine.
- Float UI
    image
- Toggle UI
    image
- Toolbar UI
    image

</details>

<details>
<summary><strong>Mobile instructions</strong></summary>

As the OSK will be loading at a later time than it used to, we need to ensure that this doesn't cause obvious blank-keyboard bugs.

It would be wise to ensure that the keyboard loads correctly and consistently in both portrait and landscape orientations (issue #5252 not withstanding) within the apps and as system keyboard.
</details>

### Explicit test list

Using `web/testing/desktop-ui` ("Desktop UI module testing") and its subpages:

- TEST_FLOAT_LOAD:  Confirm that KMW loads properly and displays the Float UI (`kmwuifloat.ts`) when the OSK is visible.
    image
- TEST_FLOAT_OSK_TOGGLE: With the Float UI active and an active keyboard (and visible OSK), repeatedly toggle the OSK on and off.
    - Note the 'toggle' button on the right-hand side of this image:  image
- TEST_FLOAT_OSK_LANGUAGE_SWAP:  With the Float UI active and an active keyboard, change the language.
    - (Possibly) after a slight delay, the OSK for the newly-selected keyboard should be displayed and the text area should still be selected.
    - Type any key (with your hardware keyboard) to confirm that the text area is indeed still active.
    - Repeat all previous steps twice, verifying that the OSK displays each time.
    - Now, toggle the OSK off.
    - Repeat the first four steps once (up to the previous "repeat" step), but this time verify that the OSK is not displayed afterward.

- TEST_TOOLBAR_LOAD:
    1. Confirm that KMW loads properly and displays the Toolbar UI (`kmwuitoolbar.ts`).
            image
    2. Use the Toolbar UI to select `sil_euro_latin` and confirm the keyboard swaps properly.
        - Most easily done through the "Europe" region, then the "English" language.
- TEST_TOOLBAR_OSK_TOGGLE:  With the Toolbar UI active and an active keyboard (and visible OSK), repeatedly toggle the OSK on and off.
    - Note the 'toggle' button on the right-hand side of this image:  image

- TEST_TOOLBAR_OSK_LANGUAGE_SWAP:  With the Toolbar UI active and an active keyboard, change the language to French.
    - (Possibly) after a slight delay, the OSK for the newly-selected keyboard should be displayed and the text area should still be selected.
    - Type any key (with your hardware keyboard) to confirm that the text area is indeed still active.
    - Use the drop-down beside French to select a different keyboard for the same language.
    - Repeat these steps _for the Polish language_ instead, verifying that the OSK displays each time.
    - Now, toggle the OSK off.
    - Repeat all steps, but this time verify that the OSK is not displayed afterward at any point.

### CODE REVIEW

File: web/source/kmwuifloat.ts [hunk 1]
```diff
@@ -33,131 +33,134 @@ if(!window['keyman']['ui']['name']) {
...
-     */     
+     */
     ui.toggleOSK = function()
     {
-      let osk = keymanweb.osk;
-      keymanweb['focusLastActiveElement']();
-      if(osk && osk['show']) 
+      keymanweb['activatingUI'](true);
+      // As the Float UI uses more advanced minification, we must preserve the 'osk' field name.
+      let osk = keymanweb['osk'];
```
Comment (mcdurdin) [inline web/source/kmwuifloat.ts:61]: I ~~hate~~ strongly dislike the fragility of this 😭 but I don't see a way around it at present...

Also, I understand why you added the comment, but it seems probably unnecessary to scatter throughout the source files given we use this pattern throughout (e.g. `keymanweb['activatingUI']`...). I think that advice should go in a coding style guide for the web code, perhaps in the [wiki style guide](https://github.com/keymanapp/keyman/wiki/Keyman-Code-Style-Guide) for now?

File: web/source/kmwuitoolbar.ts [hunk 2]
```diff
@@ -253,6 +253,7 @@ if(!window['keyman']['ui']['name']) {
       aNode = ui.createNode('a', null, 'kmw_button_a');
       aNode.href = '#';
       aNode.onclick = ui.showOSK;
+      aNode.onmousedown = function() { keymanweb['activatingUI'](true); };
```
Comment (mcdurdin) [inline web/source/kmwuitoolbar.ts:256]:
```suggestion
      aNode.onmousedown = function() {
        keymanweb['activatingUI'](true);
      };
```

I'd like to try and not perpetuate the bad old coding styles :grin:

File: web/source/kmwuitoolbar.ts [hunk 3]
```diff
@@ -505,7 +506,7 @@ if(!window['keyman']['ui']['name']) {
         var p1=ui.ToolBar_Text['SelectKeyboardPre']+kbd['Name'],p2=ui.ToolBar_Text['SelectKeyboardSuf'];
         if(p1.toLowerCase().indexOf(p2.toLowerCase()) < 0) p1=p1+' '+ p2;
         aNode.title = p1;
-        aNode.onclick = function(event) { return ui.selectLanguage(event, lang) };
+        aNode.onclick = function(event) { return ui.selectLanguage(event, lang) }; // FINDME
```
Comment (mcdurdin) [inline web/source/kmwuitoolbar.ts:509]: What does `// FINDME` mean? 🤣 
```suggestion
        aNode.onclick = function(event) {
          return ui.selectLanguage(event, lang);
        };
```

File: web/source/kmwuitoolbar.ts [hunk 4]
```diff
@@ -751,7 +754,9 @@ if(!window['keyman']['ui']['name']) {
      **/
     ui.showOSK = function(event)
     {
-      let osk = keymanweb.osk;
+      keymanweb['activatingUI'](true);
+      // As the Toolbar UI uses more advanced minification, we must preserve the 'osk' field name.
+      let osk = keymanweb['osk'];
 
       if(!osk) {
         return;
```
Comment (mcdurdin) [inline web/source/kmwuitoolbar.ts:762]: In this situation, `keymanweb.activatingUI(false)` will never be called.

**Review by mcdurdin — APPROVED** [review]
Some minor tweaks but LGTM!

**Comment by bharanidharanj** [comment]
* **TEST_FLOAT_LOAD (PASSED):** After clicking the text Area the OSK is visible in Float UI 
* **TEST_FLOAT_OSK_TOGGLE (PASSED):** When the toggle button is enabled, the OSK is visible. When the toggle button is disabled the OSK disappears. 
* **TEST_FLOAT_OSK_LANGUAGE_SWAP (PASSED):** This is working as expected. 
* **TEST_TOOLBAR_LOAD (PASSED):** This is working as expected. 
 

* **TEST_TOOLBAR_OSK_TOGGLE (PASSED):** When the toggle button is enabled, the OSK is visible. When the toggle button is disabled the OSK disappears. 

Note: It is possible for us to enable or disable the OSK by clicking inside the text area or outside of it. 

* **TEST_TOOLBAR_OSK_LANGUAGE_SWAP (PASSED):** This is working as expected. 

Note: All the above tests are passed in Desktop. (Windows 10)

**Comment by bharanidharanj** [comment]
@jahorton Tested the above PR in my Android Mobile Version 11 and I noticed that UI buttons are not visible in the Web page. 

..Portrait mode
 


..Landscape mode

File: web/source/kmwuitoolbar.ts [hunk 5]
```diff
@@ -751,9 +756,12 @@ if(!window['keyman']['ui']['name']) {
...
     {
-      let osk = keymanweb.osk;
+      keymanweb['activatingUI'](true);
+      // As the Toolbar UI uses more advanced minification, we must preserve the 'osk' field name.
+      let osk = keymanweb['osk'];
 
       if(!osk) {
+        keymanweb['activatingUI'](false);
         return;
       }
```
Comment (mcdurdin) [inline web/source/kmwuitoolbar.ts:766]:
```suggestion
      let osk = keymanweb['osk'];
      if(!osk) {
        return;
      }
      keymanweb['activatingUI'](true);
```

**Comment by jahorton** [comment]
> @jahorton Tested the above PR in my Android Mobile Version 11 and I noticed that UI buttons are not visible in the Web page.

That is correct.  The UI modules being tested only take effect on desktop form-factor devices.  Touch form-factor devices instead get the globe key you see in the Android screenshot but not in the desktop ones.

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 15.0.258-beta](https://keyman.com/downloads/releases/beta/15.0.258)

**Pull request merged**